zerver/tests/test_signup.py

```diff
@@ -1470,6 +1467,38 @@ def test_signup_already_active(self) -> None:
         result = self.client_get(result.url)
         self.assert_in_response("You've already registered", result)
 
+    def test_signup_with_used_email(self) -> None:
+        """
+        Check if signing up with an email used in another realm succeeds.
+        """
+        email = self.example_email('hamlet')
+        password = "newpassword"
+        realm = get_realm('lear')
+        do_set_realm_property(realm, 'default_language', u"de")
+
+        result = self.client_post('/accounts/home/', {'email': email}, subdomain="lear")
+        self.assertEqual(result.status_code, 302)
+        self.assertTrue(result["Location"].endswith(
+            "/accounts/send_confirm/%s" % (email,)))
+        result = self.client_get(result["Location"], subdomain="lear")
+        self.assert_in_response("Check your email so we can get started.", result)
+
+        # Visit the confirmation link.
+        confirmation_url = self.get_confirmation_url_from_outbox(email)
+        result = self.client_get(confirmation_url, subdomain="lear")
+        self.assertEqual(result.status_code, 200)
+
+        # Pick a password and agree to the ToS.
+        result = self.submit_reg_form_for_user(email, password, subdomain="lear")
+        self.assertEqual(result.status_code, 302)
+
+        user_profile = get_user(self.example_email("hamlet"), realm)
+        self.assertEqual(user_profile.default_language, realm.default_language)
+        from django.core.mail import outbox
+        outbox.pop()
+
+        self.assertEqual(UserProfile.objects.filter(email=email).count(), 2)
```

Comment:
This is the real payload of this test :) -- everything other than this and a couple of lines for "register with hamlet's email, on realm lear", is boilerplate that just obscures what's happening and should be minimized.

Reply:
Yeah, sorta; most of that content is common to most of our full new-user registration tests.  I've been meaning to figure out a way to use common code for most of it, but it's hard work, since a lot of these tests' purpose is to do something funny in the middle.

Reply:
Yeah -- I'm tempted to make a follow-up project of going in and refactoring a lot of these to use common code. I think there are some good patterns for that, but the next thing to do is to try out those patterns here and see.

Reply:
Given that we want this feature in production today, I think we should defer unnecessary test cleanup for this branch and do it as a follow-up project.  While I very much agree we want to clean up the patterns for our registration tests, I think it's a bad tradeoff to rabbit-hole on that (since I think fixing the registration test patterns overall would be like a day's work).

Reply:
Yep, I totally agree:

> ... make a follow-up project of ...